src/allmydata/storage/http_server.py

```diff
@@ -527,27 +544,47 @@ def read_data(offset: int, length: int) -> bytes:
             # https://tahoe-lafs.org/trac/tahoe-lafs/ticket/3861
             raise _HTTPError(http.NOT_ACCEPTABLE)
 
-    def _read_encoded(self, request, schema: Schema) -> Any:
+    def _read_encoded(
+        self, request, schema: Schema, max_size: int = 1024 * 1024
+    ) -> Any:
         """
         Read encoded request body data, decoding it with CBOR by default.
 
-        Somewhat arbitrarily, limit body size to 1MB; this may be too low, we
-        may want to customize per query type, but this is the starting point
-        for now.
+        Somewhat arbitrarily, limit body size to 1MiB by default.
         """
         content_type = get_content_type(request.requestHeaders)
-        if content_type == CBOR_MIME_TYPE:
-            # Read 1 byte more than 1MB. We expect length to be 1MB or
-            # less; if it's more assume it's not a legitimate message.
-            message = request.content.read(1024 * 1024 + 1)
-            if len(message) > 1024 * 1024:
-                raise _HTTPError(http.REQUEST_ENTITY_TOO_LARGE)
-            schema.validate_cbor(message)
-            result = loads(message)
-            return result
-        else:
+        if content_type != CBOR_MIME_TYPE:
             raise _HTTPError(http.UNSUPPORTED_MEDIA_TYPE)
 
+        # Make sure it's not too large:
+        request.content.seek(SEEK_END, 0)
+        if request.content.tell() > max_size:
+            raise _HTTPError(http.REQUEST_ENTITY_TOO_LARGE)
+        request.content.seek(SEEK_SET, 0)
+
+        # We don't want to load the whole message into memory, cause it might
+        # be quite large. The CDDL validator takes a read-only bytes-like
+        # thing. Luckily, for large request bodies twisted.web will buffer the
+        # data in a file, so we can use mmap() to get a memory view. The CDDL
+        # validator will not make a copy, so it won't increase memory usage
+        # beyond that.
+        try:
+            fd = request.content.fileno()
+        except (ValueError, OSError):
+            fd = -1
+        if fd > 0 and not PYCDDL_BYTES_ONLY:
+            # It's a file, so we can use mmap() to save memory.
+            message = mmap.mmap(fd, 0, access=mmap.ACCESS_READ)
```

Comment:
Looking at:

* https://gitlab.com/tahoe-lafs/pycddl/-/blob/main/src/lib.rs#L78
* https://docs.rs/ciborium/latest/ciborium/de/fn.from_reader.html
* https://docs.rs/ciborium-io/0.2.0/ciborium_io/trait.Read.html

I'm not sure why we need an mmap and not just a regular file here.  I guess "it's more work"?  Which is a fine reason not to do it that way.  So nothing actionable here.

Reply:
Originally I was using the higher-level CDDL API, which does not take a reader, so partially it's just that history. And yeah it'd be more work to make Rust's Read trait work  with Python files, pyo3 doesn't have that.

Also my hope is to eventually have a better API on CDDL side, https://github.com/anweiss/cddl/issues/167, although that may be tricky and/or never happen.